# Exercise 2.A Fundamentals of pandas

In [1]:
import numpy as np
import pandas as pd

## Series

In [2]:
aa = [1, 2, 3, 4]
ab = (5, 6, 7, 8)
ac = {9, 10, 11, 12}
ad = {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}

In [3]:
ser1 = pd.Series(aa)
ser2 = pd.Series(ab)
ser3 = pd.Series(list(ac))
ser4 = pd.Series(ad)

print("ser1:\n", ser1)
print("\nser2:\n", ser2)
print("\nser3:\n", ser3)
print("\nser4:\n", ser4)

ser1:
 0    1
1    2
2    3
3    4
dtype: int64

ser2:
 0    5
1    6
2    7
3    8
dtype: int64

ser3:
 0     9
1    10
2    11
3    12
dtype: int64

ser4:
 first     13
second    14
third     15
fourth    16
dtype: int64


In [4]:
ser5 = pd.Series(17, 18, 19, 20)
print(ser5)

# Attempt to create ser5 with literal values 17, 18, 19, 20
# This raises a TypeError because pd.Series() expects a single iterable, can not pass multiple positional arguments.
# TypeError: Index(...) must be called with a collection of some kind, 18 was passed

TypeError: Index(...) must be called with a collection of some kind, 18 was passed

## DataFrames from Series

In [5]:
df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


In [6]:
df1.insert(1, 'ser2', ser2)
df1.insert(2, 'ser3', ser3)
df1

,0,ser2,ser3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [7]:
df1.insert(3, 'ser4', ser4)
df1
# ser4 was created from a dictionary (ad), so it has string index labels ('first', 'second', 'third', 'fourth') instead of integer positions (0-3).
# But df1 index is integer-based so pandas cannot align ser4 string labels to df1 integer index, resulting in NaN for all ser4 values.

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


In [8]:
df1.loc[4] = [5, 9, 12, 0]
df1

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [26]:
df2 = pd.read_csv(r'C:\Users\malaikatariq\DataAnalytics\week-07\sakila_example.csv')
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [27]:
df2.head()

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [29]:
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [30]:
df2.info

<bound method DataFrame.info of                  title                                        description  \
0     ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
1     ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
2     ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
3     ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
4     ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
..                 ...                                                ...   
257  ATTRACTION NEWTON  A Astounding Panorama of a Composer And a Fris...   
258  ATTRACTION NEWTON  A Astounding Panorama of a Composer And a Fris...   
259        AUTUMN CROW  A Beautiful Tale of a Dentist And a Mad Cow wh...   
260        AUTUMN CROW  A Beautiful Tale of a Dentist And a Mad Cow wh...   
261        AUTUMN CROW  A Beautiful Tale of a Dentist And a Mad Cow wh...   

        category language        actor_name

**Facts from df2.info():**
- We can see the total number of rows and columns in the dataset.
- We can see the data type of each column.
- We can identify which columns have missing values.
- We can estimate memory usage of the DataFrame.

In [31]:
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


**Insights from df2.describe():**

Useful: 
- Count tells us all 262 rows have no missing values in any column
- Unique tells us how many distinct values exist
- Top tells us the most frequent value in each column
- Freq tells us how often the top value appears

Not Useful:
- Language is not useful to analyze since all 262 rows have the exact same value
- The 25th, 50th, and 75th percentiles are absent entirely so we can not detect outliers or understand data distribution

In [34]:
df3 = pd.read_csv(r'C:\Users\malaikatariq\DataAnalytics\week-07\northwind_example.csv')
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [35]:
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


**Insights from df3.describe():**

- All 500 rows are complete with no missing values
- UnitPrice ranges from 2 to 210.80, mean is $23.37
- Quantity ranges from 1 to 120, mean is 24 units
- High std of 27.33 in UnitPrice suggests possible outliers

**How it differs from df2 and why:**

- df3 has numeric columns so describe() shows mean, std, min, max, and percentiles whereas, 
- df2 had text columns so describe() only showed count, unique, top, freq
- df3 summarization is more useful for spotting outliers and understanding data distribution whereas,
- df2 summarization was limited and only useful for frequency analysis

## Coding a DataFrame

In [38]:
test_data = {
    'ID': [1, 2, 3, 4, 5, 6, 7, 8], 
    'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18],
    'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
    'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]
}

In [39]:
df4 = pd.DataFrame(test_data)
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [40]:
df4.to_csv('test_data.csv', index=False)
print("df4 exported to test_data.csv")

df4 exported to test_data.csv
